lib/Conversion/TorchToTosa/TorchToTosa.cpp

```diff
@@ -5446,8 +5446,89 @@ LogicalResult ConvertAtenOp<AtenEmptyMemoryFormatOp>::matchAndRewrite(
 
     rewriter.replaceOpWithNewOp<tosa::ConstOp>(op, resultType, emptyVal);
     return success();
+}
+
+// Calculates the min and max value of op that are observable.
+static LogicalResult inferNecessaryRange(Operation *op,
+                                         ConversionPatternRewriter &rewriter,
+                                         int64_t *min, int64_t *max) {
+    if (!op->hasOneUse()) {
+      return rewriter.notifyMatchFailure(
+          op, "can only convert round used by a single clamp");
+    }
+    Operation *user = *op->user_begin();
+
+    if (auto clamp = dyn_cast<AtenClampOp>(user)) {
+      if (!matchPattern(clamp.getMin(), m_TorchConstantInt(min)) ||
+          !matchPattern(clamp.getMax(), m_TorchConstantInt(max))) {
+        return rewriter.notifyMatchFailure(
+            op, "Clamp needs constant integer bounds");
+      }
+      return success();
+    }
+
+    if (auto add = dyn_cast<AtenAddScalarOp>(user)) {
+      int64_t alpha;
+      if (!matchPattern(add.getAlpha(), m_TorchConstantInt(&alpha)) ||
+          alpha != 1) {
+        return rewriter.notifyMatchFailure(
+            op, "Only add with alpha = 1 is handled");
+      }
+
+      int64_t summand;
+      if (!matchPattern(add.getOther(), m_TorchConstantInt(&summand))) {
+        return rewriter.notifyMatchFailure(
+            op, "Add needs constant integer second operand");
+      }
+
+      if (failed(inferNecessaryRange(add, rewriter, min, max)))
+        return failure();
+
+      // The IR looks like
+      //   a = x + summand
+      //   y = clamp a, min, max
+      // so x must have a dynamic range of [min - summand, max - summand] to
+      // preserve the output value.
+      max -= summand;
+      min -= summand;
+
+      return success();
+    }
+
+    return success();
+}
+
+// aten.round has no exact correspondence in tosa due to half-to-even
+// rounding. We can represent it as cast-to-float(cast-to-int(x)) if we can
+// find an integer type that is large enough.
+// Here we match patterns of the form clamp(round(x), min, max) where min and
+// max are within the range of an i8.
+template <>
+LogicalResult ConvertAtenOp<AtenRoundOp>::matchAndRewrite(
+    AtenRoundOp op, OpAdaptor adaptor,
+    ConversionPatternRewriter &rewriter) const {
+
+  auto outputTy = getTypeConverter()
+                      ->convertType(op.getType())
+                      .dyn_cast<RankedTensorType>();
+
```

Comment:
I think we need to constrain the type to `fp16_t`, `bf16_t` and `fp32_t`. The others do not have the expected semantics when using the tosa cast, and `AtenRound` does not seem to have any type restrictions.

Reply:
Good catch! Would be better if torch-mlir would allow lit testing for all those cases...